packages/assets-controllers/src/TokenDetectionController.test.ts

```diff
@@ -523,73 +489,44 @@ describe('TokenDetectionController', () => {
-      },
-      {
-        disabled: false,
-        isDetectionEnabledForNetwork: true,
-        isDetectionEnabledFromPreferences: false,
-        selectedAddress: '0x1',
-      },
-    );
+    tokenDetection = new TokenDetectionController({
+      chainId: ChainId.mainnet,
+      selectedAddress: '0x1',
+      onPreferencesStateChange,
+      getBalancesInSingleCall: getBalancesInSingleCallMock,
+      addDetectedTokens: stub,
+      getTokensState: () => tokensController.state,
+      getPreferencesState: () => preferences.state,
+      messenger: buildTokenDetectionControllerMessenger(controllerMessenger),
+    });
+    await tokenDetection.start();
 
     // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    await preferencesStateChangeListener!({
+    preferencesStateChangeListener!({
       selectedAddress: '0x1',
       useTokenDetection: true,
     });
-
     expect(getBalancesInSingleCallMock.called).toBe(true);
   });
 
-  it('should call getBalancesInSingleCall if onNetworkStateChange is called with a chainId that supports token detection and is changed', async () => {
+  it('should call getBalancesInSingleCall if network is changed to a chainId that supports token detection', async () => {
     const stub = sinon.stub();
     const getBalancesInSingleCallMock = sinon.stub();
-    let networkStateChangeListener: (state: any) => void;
-    const onNetworkStateChange = sinon.stub().callsFake((listener) => {
-      networkStateChangeListener = listener;
-    });
-    tokenDetection = new TokenDetectionController(
-      {
-        onNetworkStateChange,
-        onTokenListStateChange: stub,
-        onPreferencesStateChange: stub,
-        getBalancesInSingleCall: getBalancesInSingleCallMock,
-        addDetectedTokens: stub,
-        getTokensState: () => tokensController.state,
-        getTokenListState: () => tokenList.state,
-        getNetworkState: () => defaultNetworkState,
-        getPreferencesState: () => preferences.state,
-        getNetworkClientById: jest.fn(),
-      },
-      {
-        disabled: false,
-        isDetectionEnabledFromPreferences: true,
-        chainId: SupportedTokenDetectionNetworks.polygon,
-        isDetectionEnabledForNetwork: true,
-        selectedAddress: '0x1',
-      },
-    );
-
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    await networkStateChangeListener!({
-      providerConfig: { chainId: ChainId.mainnet },
+    tokenDetection = new TokenDetectionController({
+      chainId: SupportedTokenDetectionNetworks.polygon,
+      selectedAddress: '0x1',
+      onPreferencesStateChange: stub,
+      getBalancesInSingleCall: getBalancesInSingleCallMock,
+      addDetectedTokens: stub,
+      getTokensState: () => tokensController.state,
+      getPreferencesState: () => preferences.state,
+      messenger: buildTokenDetectionControllerMessenger(controllerMessenger),
     });
+    await tokenDetection.start();
```

Comment:
Same as previous comment: the network changing should cause state updates.

Reply:
This one's proving to be tricky. Not only is `detectTokens()` not being triggered, it's also unclear whether the `networkDidChange` listener is ever being run in any of the tests. Moving this ticket to blocked while I investigate.

Reply:
Fixed this here: https://github.com/MetaMask/core/pull/3609/commits/71d9967d3559111d528bf279d593167ed8e82409

The issue was that I never added a publish call for a `networkDidChange` event in `changeNetwork`.

Reply:
This fix led to another test failing: `should not call getBalancesInSingleCall after stopping polling, and then switching between networks that support token detection`

I fixed this by restoring the `disabled` class field to prevent `detectTokens()` from running passively if `stop()` had been called. The `disabled` lock is lifted iff `start()` is called: https://github.com/MetaMask/core/pull/3609/commits/0dda603252cebeee328a7b409eaf35c863e4f418.

Given the discussion around removing `disabled` in V2 controllers, is there maybe a different implementation or naming that might be better?

Reply:
Good questions! I decided to move this to Slack since we're already having an active discussion around this.

Reply:
Maybe that discussion doesn't have to block this PR, though, if the purpose is just to move to BaseControllerV2. We can keep the controller as it is and then come back and change how `disabled` works if we need to.

I will re-review this PR tomorrow with this in mind.